File: packages/ts/spaces/src/identity-link.ts
/**
 * Pure-identity link tokens — a shareable public link that carries ONLY an owner's
 * identity (userId + pseudo + Ed25519 pubkey + KEM pubkey), with no credential or
 * capability embedded.
 *
 * Unlike `SpaceInviteLinkToken` / `NodeInviteLinkToken` — which both embed an
 * ephemeral private key + cap-cert and are therefore BEARER CREDENTIALS — an
 * `IdentityLink` is safe to publish openly or embed in a client app. Its only
 * trust anchor is the `ownerId ↔ edPub` derivation binding (verified offline via
 * `session.userIdFromEdPub`, hardened by a live profile key cross-check when reachable).
 *
 * Primary use: the `createResourceRequest` / `scanResourceRequests` flow in
 * `resource-requests.ts` — a requester holds only the owner's identity link
 * and delivers a sealed resource-creation request to the owner's inbox. The owner
 * decides whether to accept or reject; no authority is delegated until then.
 */
import { readProfile } from "./client.js"
import { encodeLinkFragment, decodeLinkFragment } from "@drakkar.software/starfish-protocol"
import { hexToBytes } from "@drakkar.software/starfish-keyring"
import { ed25519 } from "@noble/curves/ed25519.js"
import { signKemSig } from "./request-verify.js"
import type { Session } from "./session.js"

// ── Regex constants ───────────────────────────────────────────────────────────

const HEX64 = /^[0-9a-f]{64}$/i
const ED_PUB_HEX_RE = HEX64
const KEM_PUB_HEX_RE = HEX64
const KEM_SIG_HEX_RE = /^[0-9a-f]{128}$/i
const USER_ID_HEX_RE = /^[0-9a-f]{32}$/i

// ── Token shape ───────────────────────────────────────────────────────────────

/**
 * Portable public identity — the only content of an identity link.
 * `pseudo` is a display hint only; KEYS are the trust anchor; `ownerId` is
 * deterministically bound to `edPub` by `userIdFromEdPub` (verified offline).
 * `kemSig` is an Ed25519 signature of kemPub bytes by edPriv, binding kemPub
 * to edPub in a way that can be verified offline without a server round-trip.
 */
export interface IdentityLink {
  v: 2
  ownerId: string
  pseudo: string
  edPub: string
  kemPub: string
  kemSig: string // Ed25519 sig of hexToBytes(kemPub) by edPriv
}

const MALFORMED = "That identity link is malformed or incomplete."

// ── Offline trust anchor ──────────────────────────────────────────────────────

/**
 * Verify the hard, OFFLINE binding:
 *   1. `token.ownerId === sha256(token.edPub)[0:32]`
 *   2. `token.kemSig` is a valid Ed25519 signature of `kemPub` bytes by `edPub`
 * Call this before rendering anything about the owner and before sending any request.
 */
export async function verifyIdentityLinkBinding(token: IdentityLink, session: Session): Promise<boolean> {
  const ownerIdMatch = (await session.userIdFromEdPub(token.edPub)) === token.ownerId
  if (!ownerIdMatch) return false
  try {
    return ed25519.verify(hexToBytes(token.kemSig), hexToBytes(token.kemPub), hexToBytes(token.edPub))
  } catch {
    return false
  }
}

// ── Encode / decode ───────────────────────────────────────────────────────────

/**
 * Pack an identity link into a URL: `<origin>/<path>#<base64url(token)>`.
 * The token rides in the URL fragment.
 */
export function encodeIdentityLink(origin: string, path: string, token: IdentityLink): string {
  return encodeLinkFragment(origin, path, token)
}

/**
 * Decode + shape-check a `#…` fragment (with or without the leading `#`).
 * Synchronous shape validation only — the `ownerId ↔ edPub` binding and kemSig
 * are verified asynchronously via {@link verifyIdentityLinkBinding}.
 * Rejects v:1 links (clean break — callers must re-publish with v:2).
 */
export function decodeIdentityLink(fragment: string): IdentityLink {
  const raw = decodeLinkFragment<Partial<IdentityLink>>(
    fragment,
    (tok: unknown) => {
      const t = tok as Partial<IdentityLink>
      return !!t &&
        t.v === 2 &&
        typeof t.ownerId === "string" &&
        USER_ID_HEX_RE.test(t.ownerId) &&
        typeof t.edPub === "string" &&
        ED_PUB_HEX_RE.test(t.edPub) &&
        typeof t.kemPub === "string" &&
        KEM_PUB_HEX_RE.test(t.kemPub) &&
        typeof t.kemSig === "string" &&
        KEM_SIG_HEX_RE.test(t.kemSig)
        ? t
        : null
    },
    MALFORMED,
  )
  return {
    v: 2,
    ownerId: raw.ownerId!,
    pseudo: typeof raw.pseudo === "string" ? raw.pseudo : "",
    edPub: raw.edPub!,
    kemPub: raw.kemPub!,
    kemSig: raw.kemSig!,
  }
}

// ── Own link ──────────────────────────────────────────────────────────────────

/**
 * Build this account's own identity link — derivable on ANY device, always the same.
 * The root device reads published keys straight from the session; a paired device
 * reads them from the (cached) public profile, like any peer would.
 *
 * Returns `null` only if the profile keys have not been published yet (brand-new
 * identity that has never synced). Call `ensureProfileKeys` first if needed.
 *
 * @param session  The current session.
 * @param origin   Web app base URL (e.g. `https://app.example.com`).
 * @param path     Route fragment (e.g. `request` → `…/request#token`). No leading `/`.
 */
export async function myIdentityLink(
  session: Session,
  origin: string,
  path: string,
): Promise<string | null> {
  // Root device: keys are already on the session — compute kemSig locally.
  if (session.ownerEdPub === session.keys.edPub) {
    const kemSig = signKemSig(session.keys)
    return encodeIdentityLink(origin, path, {
      v: 2,
      ownerId: session.userId,
      pseudo: session.name,
      edPub: session.keys.edPub,
      kemPub: session.keys.kemPub,
      kemSig,
    })
  }
  // Paired device: read published keys + kemSig from the profile.
  const profile = await readProfile(session.userId, { baseUrl: session.baseUrl, layout: session.layout })
  if (!profile.edPub || !profile.kemPub || !profile.kemSig) return null
  return encodeIdentityLink(origin, path, {
    v: 2,
    ownerId: session.userId,
    pseudo: session.name,
    edPub: profile.edPub,
    kemPub: profile.kemPub,
    kemSig: profile.kemSig,
  })
}

// ── Verify a received token against the live profile ─────────────────────────

/**
 * Cross-check a decoded token against the owner's published profile when the server
 * is reachable. Throws if the live profile has DIFFERENT keys than the token.
 * Succeeds silently when the profile is unreachable.
 */
export async function verifyIdentityLinkKeys(token: IdentityLink, session: Session): Promise<void> {
  const profile = await readProfile(token.ownerId, { baseUrl: session.baseUrl, layout: session.layout }).catch(() => null)
  if (!profile) return
  if (
    (profile.edPub && profile.edPub !== token.edPub) ||
    (profile.kemPub && profile.kemPub !== token.kemPub)
  ) {
    throw new Error("This identity link doesn't match the owner's published identity keys.")
  }
}

File: packages/python/spaces/starfish_spaces/identity_link.py
"""Pure-identity link tokens (v:2) — safe to publish openly.

Unlike space/node invite links (bearer credentials), an identity link carries
ONLY the owner's identity (``ownerId`` + ``edPub`` + ``kemPub`` + ``kemSig``).
The only trust anchor is the ``ownerId ↔ edPub`` derivation binding.

Primary use: the ``create_resource_request`` / ``scan_resource_requests`` flow in
:mod:`starfish_spaces.resource_requests`.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Callable, Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from starfish_protocol.encoding import decode_link_fragment, encode_link_fragment

from starfish_spaces.request_verify import sign_kem_sig

if TYPE_CHECKING:
    from starfish_spaces.session import Session

# ── Regex constants (mirror TS; use \Z anchor + [0-9a-f] char class) ─────────

_HEX64 = re.compile(r"^[0-9a-f]{64}\Z", re.IGNORECASE)
_ED_PUB_HEX_RE = _HEX64
_KEM_PUB_HEX_RE = _HEX64
_KEM_SIG_HEX_RE = re.compile(r"^[0-9a-f]{128}\Z", re.IGNORECASE)
_USER_ID_HEX_RE = re.compile(r"^[0-9a-f]{32}\Z", re.IGNORECASE)

_MALFORMED = "That identity link is malformed or incomplete."


# ── Token shape ───────────────────────────────────────────────────────────────


@dataclass
class IdentityLink:
    """Portable public identity — the only content of an identity link.

    ``kemSig`` is an Ed25519 signature of ``hexToBytes(kemPub)`` by ``edPriv``,
    binding ``kemPub`` to ``edPub`` in a way that can be verified offline.
    """

    v: int  # always 2
    owner_id: str
    pseudo: str
    ed_pub: str
    kem_pub: str
    kem_sig: str


# ── Offline trust anchor ──────────────────────────────────────────────────────


async def verify_identity_link_binding(
    token: IdentityLink,
    session: "Session",
) -> bool:
    """Verify the hard OFFLINE binding.

    1. ``token.owner_id == sha256(token.ed_pub)[:16].hex()``
    2. ``token.kem_sig`` is a valid Ed25519 signature of ``kemPub`` by ``edPub``.

    Call before rendering anything about the owner or sending any request.
    """
    expected = await session.user_id_from_ed_pub(token.ed_pub)
    if expected != token.owner_id:
        return False
    try:
        from starfish_spaces.request_verify import verify_kem_sig
        return verify_kem_sig(token.ed_pub, token.kem_pub, token.kem_sig)
    except Exception:
        return False


# ── Encode / decode ───────────────────────────────────────────────────────────


def encode_identity_link(origin: str, path: str, token: IdentityLink) -> str:
    """Pack an identity link into a URL: ``<origin>/<path>#<base64url(token)>``."""
    payload = {
        "v": token.v,
        "ownerId": token.owner_id,
        "pseudo": token.pseudo,
        "edPub": token.ed_pub,
        "kemPub": token.kem_pub,
        "kemSig": token.kem_sig,
    }
    return encode_link_fragment(origin, path, payload)


def _validate_identity_link(tok: Any) -> Optional[dict[str, Any]]:
    """Validator for :func:`decode_link_fragment` — returns token dict or ``None``."""
    if not isinstance(tok, dict):
        return None
    if tok.get("v") != 2:
        return None
    owner_id = tok.get("ownerId", "")
    ed_pub = tok.get("edPub", "")
    kem_pub = tok.get("kemPub", "")
    kem_sig = tok.get("kemSig", "")
    if not (
        isinstance(owner_id, str) and _USER_ID_HEX_RE.match(owner_id)
        and isinstance(ed_pub, str) and _ED_PUB_HEX_RE.match(ed_pub)
        and isinstance(kem_pub, str) and _KEM_PUB_HEX_RE.match(kem_pub)
        and isinstance(kem_sig, str) and _KEM_SIG_HEX_RE.match(kem_sig)
    ):
        return None
    return tok


def decode_identity_link(fragment: str) -> IdentityLink:
    """Decode + shape-check a ``#…`` fragment.

    Synchronous shape validation only — the ``ownerId ↔ edPub`` binding is
    verified asynchronously via :func:`verify_identity_link_binding`.
    Rejects v:1 links (callers must re-publish with v:2).

    Raises:
        ValueError: if the fragment is malformed or fails shape-check.
    """
    raw: dict[str, Any] = decode_link_fragment(fragment, _validate_identity_link, _MALFORMED)
    return IdentityLink(
        v=2,
        owner_id=raw["ownerId"],
        pseudo=raw.get("pseudo") or "",
        ed_pub=raw["edPub"],
        kem_pub=raw["kemPub"],
        kem_sig=raw["kemSig"],
    )


# ── Own identity link ─────────────────────────────────────────────────────────


async def my_identity_link(
    session: "Session",
    origin: str,
    path: str,
) -> Optional[str]:
    """Build this account's own identity link (derivable on any device).

    Returns ``None`` only if the profile keys have not been published yet.
    """
    # Root device: keys are already on the session — compute kemSig locally.
    if session.owner_ed_pub == session.keys["edPub"]:
        kem_sig = sign_kem_sig(session.keys["kemPub"], session.keys["edPriv"])
        return encode_identity_link(
            origin, path,
            IdentityLink(
                v=2,
                owner_id=session.user_id,
                pseudo=session.name,
                ed_pub=session.keys["edPub"],
                kem_pub=session.keys["kemPub"],
                kem_sig=kem_sig,
            ),
        )
    # Paired device: read published keys from the profile.
    from starfish_spaces.client import read_profile
    profile = await read_profile(session.account_client, session.user_id, session.layout)
    if not (profile.get("edPub") and profile.get("kemPub") and profile.get("kemSig")):
        return None
    return encode_identity_link(
        origin, path,
        IdentityLink(
            v=2,
            owner_id=session.user_id,
            pseudo=session.name,
            ed_pub=profile["edPub"],
            kem_pub=profile["kemPub"],
            kem_sig=profile["kemSig"],
        ),
    )


# ── Live cross-check ──────────────────────────────────────────────────────────


async def verify_identity_link_keys(
    token: IdentityLink,
    session: "Session",
) -> None:
    """Cross-check a decoded token against the owner's published profile.

    Throws if the profile has DIFFERENT keys than the token.
    Succeeds silently when the profile is unreachable.
    """
    from starfish_spaces.client import make_anon_space_client, ClientOpts

    try:
        from starfish_spaces.client import read_profile
        anon_opts: ClientOpts = {"baseUrl": session.base_url, "namespace": session.namespace}  # type: ignore[misc]
        anon_client = make_anon_space_client(anon_opts)
        profile = await read_profile(anon_client, token.owner_id, session.layout)
    except Exception:
        return  # unreachable — succeed silently

    if profile.get("edPub") and profile["edPub"] != token.ed_pub:
        raise ValueError("This identity link doesn't match the owner's published identity keys.")
    if profile.get("kemPub") and profile["kemPub"] != token.kem_pub:
        raise ValueError("This identity link doesn't match the owner's published identity keys.")


__all__ = [
    "IdentityLink",
    "verify_identity_link_binding",
    "encode_identity_link",
    "decode_identity_link",
    "my_identity_link",
    "verify_identity_link_keys",
]
